internal/util/importutilv2/csv/reader.go

```diff
@@ -0,0 +1,129 @@
+package csv
+
+import (
+	"context"
+	"encoding/csv"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/milvus-io/milvus-proto/go-api/v2/schemapb"
+	"github.com/milvus-io/milvus/internal/storage"
+	"github.com/milvus-io/milvus/pkg/util/merr"
+	"github.com/milvus-io/milvus/pkg/util/typeutil"
+	"go.uber.org/atomic"
+)
+
+type Row = map[storage.FieldID]any
+
+type reader struct {
+	ctx    context.Context
+	cm     storage.ChunkManager
+	schema *schemapb.CollectionSchema
+
+	cr     *csv.Reader
+	parser RowParser
+
+	fileSize   *atomic.Int64
+	bufferSize int
+	count      int64
+	filePath   string
+}
+
+func NewReader(ctx context.Context, cm storage.ChunkManager, schema *schemapb.CollectionSchema, path string, bufferSize int, sep rune) (*reader, error) {
+	cmReader, err := cm.Reader(ctx, path)
+	if err != nil {
+		return nil, merr.WrapErrImportFailed(fmt.Sprintf("read csv file failed, path=%s, err=%s", path, err.Error()))
+	}
+	count, err := estimateReadCountPerBatch(bufferSize, schema)
+	if err != nil {
+		return nil, err
+	}
+	csvReader := csv.NewReader(cmReader)
+	csvReader.Comma = sep
+
+	header, err := csvReader.Read()
+	fmt.Printf("csv header: %v\n", strings.Join(header, "|"))
+	fmt.Printf("length: %v\n", len(header))
+	if err != nil {
+		return nil, merr.WrapErrImportFailed(fmt.Sprintf("failed to read csv header, error: %v", err))
+	}
+
+	rowParser, err := NewRowParser(schema, header)
+	if err != nil {
+		return nil, err
+	}
+	return &reader{
+		ctx:        ctx,
+		cm:         cm,
+		schema:     schema,
+		cr:         csvReader,
+		parser:     rowParser,
+		fileSize:   atomic.NewInt64(0),
+		filePath:   path,
+		bufferSize: bufferSize,
+		count:      count,
+	}, nil
+}
+
+func (r *reader) Read() (*storage.InsertData, error) {
+	insertData, err := storage.NewInsertData(r.schema)
+	if err != nil {
+		return nil, err
+	}
+	var cnt int64 = 0
+	for {
+		value, err := r.cr.Read()
+		if err == io.EOF || len(value) == 0 {
+			break
+		}
+		row, err := r.parser.Parse(value)
+		if err != nil {
+			return nil, merr.WrapErrImportFailed(fmt.Sprintf("failed to parse row, error: %v", err))
+		}
+		err = insertData.Append(row)
+		if err != nil {
+			return nil, merr.WrapErrImportFailed(fmt.Sprintf("failed to append row, error: %v", err))
+		}
+		cnt++
+		if cnt >= r.count {
```

Comment:
If we can get actual memory size based on `insertData.GetMemorySize()`, it would be no need to estimate row size and control memory consumption based on row count.

Reply:
> If we can get actual memory size based on `insertData.GetMemorySize()`, it won't be necessary to estimate row size and control memory consumption based on row count.

@tedxu Yes, we're using `insertData.GetMemorySize()`. The idea of introducing count is to avoid frequent calls to `insertData.GetMemorySize()`.

Reply:
So, I will keep the use of cnt?